### PR TITLE
README, API Docs, and CI Checks

### PR DESCRIPTION
**Pull request by qwang98**
NOTES FOR README
Let me know if the Chiquito circuit example and integration with base Halo2 example are readable.
TODO: Replace API doc link after merged.

NOTES FOR API DOC
A few design principles I followed:
1. document it from the dev's workflow perspective, so start from the overall workflow for its integration in halo2, to building a circuit, to building a step
2. only document functions that are used by the devs, and most of them are from the DSL. there are functions in the DSLs that are only called internally, which I didn't document. there are also functions that are not in the DSLs but are called by the users regardless, e.g. the assign function for wg and fixed_gen, which I documented

### CODE REVIEW

File: API Documentation [hunk 1]
```diff
@@ -0,0 +1,740 @@
...
+- Call `chiquito2Halo2` function in Halo2 backend to convert the `ir::Circuit` object to a `ChiquitoHalo2` object
+- Integrate `ChiquitoHalo2` object into a Halo2 circuit using a boiler plate code
+
+# DSL Functions
+
+The above section describes the high level process of building and integrating a Chiquito Halo2 backend object into a Halo2 circuit. However, when building a circuit using Chiquito, the developer mostly call DSL functions to manipulate the `ast::Circuit` object, which the rest of this API documentation will focus on.
+
+Before we dive in, there are a few Chiquito-specific concepts that are not in Halo2:
+
+- Step (or super row): Chiquito provides syntax sugar for defining complex constraints and therefore sometimes requires allocating multiple Halo2 rows to build out a set of constraints. These groups of rows are called “super rows” or “steps”.
```
Comment (leolara) [inline API Documentation:22]: a super row is the rows and columns that a step instance uses in the plonkish table

File: API Documentation [hunk 2]
```diff
@@ -0,0 +1,740 @@
...
+- Signals: Halo2 uses rotations to describe constraints among different rows. Chiquito abstracts rotation away from developers by using “signals”, to mimic Circom’s user friendly features.
+    - Internal signal: only supports constraints within the same step
+    - Forward signal: supports constraints across different steps
```
Comment (leolara) [inline API Documentation:25]: forward, also constraint on the value in the next step instance

File: API Documentation [hunk 3]
```diff
@@ -0,0 +1,740 @@
...
+### **`pragma_first_step` function**
+
+```rust
+pub fn pragma_first_step(&mut self, step_type: StepTypeHandler)
+
+```
+
+**Description:**
+
+Enforce the type of the first step for the circuit using the **`StepTypeHandler`** of the desired StepType
```
Comment (leolara) [inline API Documentation:224]: I would say that enforce by adding a constraint to the circuit

File: README.md [hunk 4]
```diff
@@ -1,2 +1,490 @@
...
+        step_selector::SimpleStepSelectorBuilder, // input for constructing the compiler
+        Compiler, // compiles AST to IR
+    },
+    ir::Circuit, // IR object that the compiler compiles to
+    backend::halo2::{chiquito2Halo2, ChiquitoHalo2}, // compiles to Chiquito Halo2 backend, which can be integrated into Halo2 circuit
+    stdlib::*, // standard library
+};
+```
+
+## Example: Build zkEVM Bytecode Chiquito Circuit
```
[the remark this answers is not in the document]
Comment (leolara) [reply README.md:95]: You don't need to include all  the "use" cases, that is a rust detail

**Review by ChihChengLiang — CHANGES_REQUESTED** [review]
Hi @qwang98, great effort for putting together the README and docs.
We can leverage some existing Rust toolings to simplify them. Let me summarize my points below.
1. We should keep the README short. We should move the code examples in the README into the examples/ directory so that compilers, tests, and style checks can cover those codes.
2. We should write documentation directly on the Rust code. By using `cargo doc`, we can build beautiful documentation pages.
3. Additionally. It would be great if we could split this pull request into 3 separate ones so that we can better separate the review and merging efforts.

File: API Documentation (hunk 1 above)
Comment (qwang98) [reply API Documentation:22]: Edits:
- Step (or super row): Each "step" uses one or multiple rows and columns in the plonkish table to define constraints and witness generation. That's why "steps" are also called "super rows".

File: API Documentation (hunk 2 above)
Comment (qwang98) [reply API Documentation:25]: Edits
- Forward signal: supports constraints across different steps, including constraint on the value in the next step instance

File: API Documentation [hunk 5]
```diff
@@ -0,0 +1,740 @@
...
+
+- Step (or super row): Chiquito provides syntax sugar for defining complex constraints and therefore sometimes requires allocating multiple Halo2 rows to build out a set of constraints. These groups of rows are called “super rows” or “steps”.
+- Signals: Halo2 uses rotations to describe constraints among different rows. Chiquito abstracts rotation away from developers by using “signals”, to mimic Circom’s user friendly features.
+    - Internal signal: only supports constraints within the same step
+    - Forward signal: supports constraints across different steps
+
+DSL functions are defined on two levels:
+
+- Circuit level: define and manipulate circuit-level objects, such as forward signals, step types, fixed columns, and imported Halo2 columns.
+- Step type level: defined and manipulate step-specific objects, such as internal signals, constraints, witness generations.
```
[the remark this answers is not in the document]
Comment (qwang98) [reply API Documentation:30]: Edits:

DSL functions are defined on five different levels, with nested relationships:

- Circuit level: define and manipulate circuit-level objects, such as forward signals, step types, fixed columns, and imported Halo2 columns
    - Step type level: define and manipulate step-specific objects, such as internal signals, constraints, witness generations
        - Witness generation level: allow user-defined Turing-complete function to manipulate witness generation inputs and assign witness values
    - Fixed column generation level: allow user-defined Turing-complete function to manipulate fixed column generation inputs and assign fixed column values
    - Trace level: create the main circuit by instantiating step types; allow user-defined Turing-complete function to manipulate external trace inputs and assign input values to step type instances

File: API Documentation (hunk 3 above)
Comment (qwang98) [reply API Documentation:224]: Edits for pragma_first_step:
Enforce the type of the first step by adding a constraint to the circuit.

for pragma_last_step:
Enforce the type of the last step by adding a constraint to the circuit.

File: README.md [hunk 6]
```diff
@@ -1,2 +1,490 @@
...
+Add the following line to `Cargo.toml` of your project:
+
+```rust
+chiquito = { git = "[https://github.com/privacy-scaling-explorations/chiquito](https://github.com/privacy-scaling-explorations/chiquito)", tag = "v2023_04_24_2" }
+```
+
+Paste the following standard crate import header to the top of your circuit code file:
+
+```rust
+use chiquito::{
```
[the remark this answers is not in the document]
Comment (qwang98) [reply README.md:75]: Makes sense. The Bytecode example is a bit complicated as Leo pointed out and I'll use an existing example in the folder instead, and following your suggestions for documenting directly on the code.

File: API Documentation (hunk 1 above)
Comment (leolara) [reply API Documentation:22]: A super row is only related to the step instance, not the the step definition. A step definition has no rows, jury constraints

**Comment by qwang98** [comment]
README related updates now in pr #12
API doc related updates now in pr #13 
This pr will eventually be closed without merging

File: README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (qwang98) [reply README.md:95]: fibonacci now used in pr #12 

all code excluded from README now

now only link to code example is provided

File: API Documentation (hunk 1 above)
Comment (qwang98) [reply API Documentation:22]: fixed

**Comment by qwang98** [comment]
CI check now ready in issue #15 (per cc's request to separate into 3 pr)
Again, this pr will be eventually closed without merging

**Comment by ChihChengLiang** [comment]
Close in favor of #12 and #13

**Pull request closed without merging**